File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.py
from collections import defaultdict

def solution(phone_book):
    phone_book.sort(key = lambda x: len(x))
    phone_dict = defaultdict(int)
    
    for i in range(len(phone_book)):
        temp = ''
        for j in range(len(phone_book[i])):
            temp += phone_book[i][j]
            if phone_dict[temp]:
                return False
        phone_dict[phone_book[i]] = 1
    return True

File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    public boolean solution(String[] phone_book) {
        boolean answer = true;
        Map<String, Boolean> map = new HashMap<>();
        Arrays.sort(phone_book, (o1, o2) -> o1.length() - o2.length());
        
        
        for(String number : phone_book){
            StringBuilder sb = new StringBuilder();
            for(int j=0 ; j<number.length() ; j++){
                sb.append(number.charAt(j));
                if(map.containsKey(sb.toString()))return false;
            }
            map.put(sb.toString(), true);
        }
        return true;
    }
}